In [ ]:
import numpy as np
from sklearn.impute import KNNImputer
from sklearn.metrics import nan_euclidean_distances as dist

A continuación veremos como opera kNN mediante un ejemplo muy sencillo pero ilustrativo y haremos los calculos para verificar los valores que imputa el algoritmo

In [20]:
X = [[1, 2, np.nan], [3, 4, 3], [np.nan, 6, 5], [8, 8, 7]]
imputer = KNNImputer(n_neighbors=2)
imputer.fit_transform(X)

array([[1. , 2. , 4. ],
       [3. , 4. , 3. ],
       [5.5, 6. , 5. ],
       [8. , 8. , 7. ]])

Ahora veremos que para el primer punto los 2 vecinos más cercanos son $[3, 4, 3], [np.nan, 6, 5]$ por lo tanto para usa el promedio de ambos en la tercera coordenada para imputar el nan, es decir $(3+5)/2=4$ por lo tanto $[1, 2, np.nan]->[1, 2, 4]$. Pero ¿como reproducimos los calculos anteriores?. A continuación lo veremos, primero calculando la matriz de distancia euclideana y haciendo los cálculos de cada una de los componentes

In [19]:
dist(X,X)

array([[ 0.        ,  3.46410162,  6.92820323, 11.29158979],
       [ 3.46410162,  0.        ,  3.46410162,  7.54983444],
       [ 6.92820323,  3.46410162,  0.        ,  3.46410162],
       [11.29158979,  7.54983444,  3.46410162,  0.        ]])

Recordemos que aquí la distancia con valores faltantes se define como:
$$
d_{x_0,x_1}=\sqrt{\left(weight* dist({x_0,x_1}\right)}
$$

donde weight es el la dimensión entre el número de coordenadas no missing en los puntos, ejemplo en el punto $x_0=[1, 2, np.nan]$, $w=3/2$ entonces la distancia entre $x_0=[1, 2, np.nan]$ y $x_0=( [3, 4, 3]$ es: 
$$
\begin{align*}
d_{x_0,x_1}&=\sqrt{\left(3/2\times dist({x_0,x_1}\right)}\\
&=\sqrt{\left(3/2\times [(1-3)^2+(2-4)^2]\right)} \\
&=3.46410162
\end{align*}
$$
Ahora sí quisieramos calcular la distancia entre el punto $x_0$ y el punto $x_2=[np.nan, 6, 5]$ tendríamos que $w_{02}=3/1$ aquí el número de coordenadas presentes es 1 porque la primera coordenada de $x_0$ es nan y la segunda de $x_2$ también es nan, así que solamente nos queda una dimensión para operar
$$
\begin{align*}
d_{x_0,x_3}&=\sqrt{\left(3/1\times dist({x_0,x_1}\right)}\\
&=\sqrt{\left(3\times [(2-6)^2]\right)} \\
&=6.92820323
\end{align*}
$$

In [25]:
Y = [[np.nan], [4], [5], [7]]
imputer = KNNImputer(n_neighbors=3)
imputer.fit_transform(Y)

array([[5.33333333],
       [4.        ],
       [5.        ],
       [7.        ]])

Cuando tienes un punto en una sola dimensión dado que no puedes determinar cual es el vecino más cercano si usas Knn será equivalente a promediar sobre todos los puntos en el dataset, así knn es útil para n>=2.